daemon/daemon.go

```diff
@@ -836,11 +861,24 @@ func NewDaemon(ctx context.Context, config *config.Config, pluginStore *plugin.S
 	}
 	d.configStore.Store(cfgStore)
 
-	// TEST_INTEGRATION_USE_SNAPSHOTTER is used for integration tests only.
-	if os.Getenv("TEST_INTEGRATION_USE_SNAPSHOTTER") != "" {
-		d.usesSnapshotter = true
-	} else {
-		d.usesSnapshotter = config.Features["containerd-snapshotter"]
+	migrationThreshold := int64(-1)
+	tryGraphDriver := graphdriver.IsRegistered
+	if os.Getenv("TEST_INTEGRATION_USE_GRAPHDRIVER") != "" {
+		tryGraphDriver = func(driver string) bool {
+			if driver == "" || graphdriver.IsRegistered(driver) {
+				return true
+			}
+			log.G(ctx).WithField("driver", driver).Warn("TEST_INTEGRATION_USE_GRAPHDRIVER is set but graphdriver is not registered")
+			return false
```

Comment:
Shouldn't this be a hard failure? Usage of graphdrivers was explicitly requested (`TEST_INTEGRATION_USE_GRAPHDRIVER != ""`), but the config specifies an unknown driver.

Reply:
Updated